ECS reporter: Minimize API calls by caching task and service data

ekimekim opened the pull request:
Fixes https://github.com/weaveworks/scope/issues/2050

Due to AWS API rate limits, we need to minimize API calls as much as
possible.

Our stated objectives:
* for all displayed tasks and services to have up-to-date metadata
* for all tasks to map to services if able

My approach here:
* Tasks only contain immutable fields (that we care about). We cache tasks forever.
 We only DescribeTasks the first time we see a new task.
* We attempt to match tasks to services with what info we have. Any "referenced" services,
 ie. a service with at least one matching task, needs to be updated to refresh changing data.
* In the event that a task doesn't match any of the (updated) services, ie. a new service entirely
 needs to be found, we do a full list and detail of all services (we don't re-detail ones we just refreshed).
* To avoid unbounded memory usage, we evict tasks and services from the cache after 1 minute without use.
 This should be long enough for things like temporary failures to be glossed over.

This gives us exactly one call per task, and one call per referenced service per report, which is unavoidable to maintain fresh data. Expensive "describe all" service queries are kept to only when newly-referenced services appear, which should be rare.

We could make a few very minor improvements here, such as trying to refresh unreferenced but known services before doing a list query, or getting details one by one when "describing all" and stopping when all matches have been found, but I believe these would produce very minor, if any, gains in number of calls while having an unjustifiable effect on latency since we wouldn't be able to do requests as concurrently.

Speaking of which, this change has a minor performance impact. Even though we're now doing less calls, we can't do them as concurrently.

Old code:
```
concurrently:
	describe tasks (1 call)
	sequentially:
		list services (1 call)
		describe services (N calls concurrently)
```
Assuming full concurrency, total latency: 2 end-to-end calls

New code (worst case):
```
sequentially:
	describe tasks (1 call)
	describe services (N calls concurrently)
	list services (1 call)
	describe services (N calls concurrently)
```
Assuming full concurrency, total latency: 4 end-to-end calls

In practical terms, I don't expect this to matter.

ekimekim commented:
Tested, added a ton of debug logging and fixed a bunch of bugs.
Can confirm we do much less calls now.

In this example system, we have one service with 3 tasks across 3 machines, so one task on our example machine.
This is what the first report on startup looks like: (many extra logs elided)
```
<probe> DEBU: 2016/12/10 02:09:06.235489 Creating new ECS client
<probe> DEBU: 2016/12/10 02:09:06.324142 Describing 1 ECS tasks
<probe> DEBU: 2016/12/10 02:09:06.367489 Refreshing ECS services
<probe> DEBU: 2016/12/10 02:09:06.367542 Described 0 services in 0 calls
<probe> DEBU: 2016/12/10 02:09:06.367569 After refreshing services, 1 tasks unmatched
<probe> DEBU: 2016/12/10 02:09:06.367574 Listing ECS services
<probe> DEBU: 2016/12/10 02:09:06.376347 Listed 1 services
<probe> DEBU: 2016/12/10 02:09:06.376414 Described 1 services in 1 calls
<probe> DEBU: 2016/12/10 02:09:06.387699 Got info from ECS: 1 tasks, 1 services
```
for a total of 3 calls (1 DescribeTasks, 1 ListServices, 1 DescribeServices)

In subsequent reports:
```
<probe> DEBU: 2016/12/10 02:09:07.162438 Refreshing ECS services
<probe> DEBU: 2016/12/10 02:09:07.162537 Described 1 services in 1 calls
<probe> DEBU: 2016/12/10 02:09:07.178871 After refreshing services, 0 tasks unmatched
<probe> DEBU: 2016/12/10 02:09:07.179110 Got info from ECS: 1 tasks, 1 services
```
for a total of 1 call (DescribeServices)

2opremio commented:
> To avoid unbounded memory usage, we evict tasks and services from the cache after 1 minute without use.

Shouldn't this be size-bound instead of time-bound? Also, 1-minute seems very conservative (1 minute in scheduler time is very small).

2opremio commented:
> such as trying to refresh unreferenced but known services before doing a list query

Would this fix #2085 ? (i.e. what does _known_ mean here? previously referenced?)

2opremio commented:
> In practical terms, I don't expect this to matter.

I won't either

2opremio commented on probe/awsecs/reporter.go:
```diff
@@ -78,39 +78,51 @@ func getLabelInfo(rpt report.Report) map[string]map[string]*taskLabelInfo {
 
 // Reporter implements Tagger, Reporter
 type Reporter struct {
+	clients map[string]*ecsClient
```
Mind documenting the keys?

2opremio replied:
Maybe renaming to `clusterClients` or `clientsPerCluster` would be enough

2opremio commented on probe/awsecs/reporter.go:
```diff
@@ -78,39 +78,51 @@ func getLabelInfo(rpt report.Report) map[string]map[string]*taskLabelInfo {
+		client, ok := r.clients[cluster]
+		if !ok {
+			log.Debugf("Creating new ECS client")
+			var err error // can't use := on the next line without shadowing outer client var
```
I would remove the comment

ekimekim replied:
done

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +219,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+}
+
+// Returns a ecsInfo struct containing data needed for a report.
+func (c ecsClient) getInfo(taskARNs []string) ecsInfo {
```
Mind breaking down this function into simpler units? It got really long

2opremio commented:
Can you make a rough estimation on how many clusters/container instances/services could be run with these improvements?

Did you test this in AWS with a few clusters? 

Also, could you please add some unit tests?

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +219,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
-		servicesChan <- c.getServices()
-	}()
+// Evict entries from the caches which have not been used within the eviction interval.
+func (c ecsClient) evictOldCacheItems() {
```
I would suggest using an existing cache library. A quick search reveals we use github.com/bluele/gcache and github.com/coocood/freecache in some parts of the code

ekimekim replied:
I don't think a standard cache would be suitable. Our goal here isn't to keep a certain size budget of things cached for performance, it's to absolutely minimize the number of calls we're making by explicitly managing the cache. Things shouldn't be evicted if they're still in use - in the case of services in particular this will trigger an expensive list-and-describe-all.
The timed eviction logic is only here so we can withstand a given task flickering out of existence due to a restarting container or other temporary weirdness. Otherwise we'd be explicitly only carrying forward the items we used last round.

With this in mind, maybe the name "cache" is misleading - perhaps "known" or even just simply "tasks" and "services"?

ekimekim commented on probe/awsecs/client.go:
```diff
@@ -12,13 +14,50 @@ import (
+	client       *ecs.ECS
+	cluster      string
+	taskCache    map[string]ecsTask
+	serviceCache map[string]ecsService
```
done

ekimekim commented on probe/awsecs/reporter.go:
```diff
@@ -78,39 +78,51 @@ func getLabelInfo(rpt report.Report) map[string]map[string]*taskLabelInfo {
+}
+
+// New creates a new Reporter
+func New() Reporter {
```
done

ekimekim commented:
PTAL

2opremio commented on probe/awsecs/client.go:
```diff
@@ -31,72 +70,146 @@ func newClient(cluster string) (*ecsClient, error) {
-	lock := sync.Mutex{} // lock mediates access to results
+// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
+// with full ecsService objects being put into the cache. Closes done when finished.
+func (c ecsClient) describeServices() (chan<- string, <-chan bool) {
```
It doesn't look like you are sending anything over the done channel. I would use a `chan struct{}` instead of a `chan bool`

ekimekim replied:
Huh, cool.

2opremio commented on probe/awsecs/client.go:
```diff
@@ -31,72 +70,146 @@ func newClient(cluster string) (*ecsClient, error) {
-				for _, failure := range resp.Failures {
-					log.Warnf("Failed to describe ECS service %s, ECS service report may be incomplete: %s", failure.Arn, failure.Reason)
-				}
+		describePage := func(arns []string) {
```
To make this even easier to read I would move `describePage` out of `describeServices`. Also, instead of initializing the lock every time, wouldn't it be cleaner if it lived on the client struct?

ekimekim replied:
done

2opremio commented:
I gave it another read. I asked some questions in my first comments which remain to be answered.

ekimekim commented on probe/awsecs/client.go:
```diff
@@ -31,72 +70,146 @@ func newClient(cluster string) (*ecsClient, error) {
+
+func newECSService(service *ecs.Service) ecsService {
+	now := time.Now()
+	deploymentIDs := make([]string, 0, len(service.Deployments))
```
Yeah, I'm still in the python habit of "avoid using indices". done.

ekimekim commented:
> Can you make a rough estimation on how many clusters/container instances/services could be run with these improvements?

Taking socks shop as an (admittedly poor) example:
Socks shop has 14 services and 1 task per service, across 3 machines.
Let's take the average of 14/3 = 4.6 tasks per machine and round up to 5.

Note it's unclear if rate limits are per actual http request, or per thing being described (ie. if a describe services call with 10 services is 10x the 'cost' of a describe services with 1 service, or the same 'cost')

Old:
for each machine, list and describe all and services, and describe all tasks on that machine
= 3 machines * (1 list services + 14 describe services + 5 describe tasks)
or 3 machines * (1 list services + 2 describe service calls + 1 describe tasks call)
= 60 list or describe operations
or 12 list or describe calls

New:
first report is as per old. for subsequent reports:
for each machine, describe all services on that machine
= 3 machines * 5 describe services
or 3 machines * 1 describe service call
= 15 describe operations
or 3 describe calls

In summary: We make 1/4 the calls. But note this is somewhat of an ideal scenario, since most services aren't present on most machines.
But I'd expect we'll always save at least half of all calls, just by virtue of caching all tasks.

ekimekim commented:
rebased, resolved trivial conflict

ekimekim commented:
I'm running a test with 3 socks shop clusters in one region, to see if I hit rate limits. Leaving it 12h, which should be enough to hit any longer-term limits.

ekimekim commented:
Ran for about 15h with no rate limiting.

Can I get a PTAL % no unit tests?

2opremio commented:
Please rebase, #2094 has messed up your commits and it's hard to review.

2opremio commented:
Also, there's a pending question about size-based eviction vs time-based eviction (I would go for the first one or a combination of the two if you want to refine it).

ekimekim commented:
rebased.

The reason I went for a time-based cache is because:
* The sizes involved are small. Even if the strings end up being large, a ecsTask or ecsService struct could only be at most 100s of bytes. Supposing a worst-case scenario of 512B, we could store 100s of such tasks and services for a total cost of less than 100KB.
* In contrast, the cost of an additional AWS call is very high, since we don't know what the limits are and when you hit that limit, there's no recourse except "don't run as much stuff in one region".

Therefore, we _always_ want to prefer more caching over calls.
I added the time-based eviction as an afterthought just to ensure the cache wouldn't grow without bound in scenarios with high amounts of task churn.
Increasing the "1 minute" to "1 hour" would probably make sense, now I think of it, just to ensure we never evict until it's unreasonable to think it might ever be useful again.

The other reason I'm hesitant to use a size-based cache is because it would complicate the code which refreshes items and puts them in the cache, since we can no longer be 100% sure that items we get in step A would still be there by the time we reach step B. We would need to either introduce extra "working" maps to track things that we are currently using, or find/implement a size-based cache with an explicit "evict now" step instead of upon every addition.

2opremio commented:
I think that not setting a maximum size bound is unsafe particularly if we increase the time limit (which we should).

Otherwise the code looks good to me (module the missing tests)

2opremio commented:
> We would need to either introduce extra "working" maps to track things that we are currently using, or find/implement a size-based cache with an explicit "evict now" step instead of upon every addition.

ekimekim commented:
But what happens when the max cache size is smaller than the memory needed to store all the currently-active tasks and services? We'd end up with items disappearing before we have a chance to actually report them, and get silent data loss.

2opremio commented:
It won't happen if the size is big enough (say 10k services and tasks, if
we have more entries than that, there's something clearly wrong).

Even if it happens, it will only affect ECS views, which is preferable to
getting killed by the kernel.

In general, I think it's a good idea to set explicit size limits, because
otherwise the system will set them for you in an uncontrolled fashion.

On Dec 16, 2016 20:29, "Mike Lang" <notifications@github.com> wrote:

> But what happens when the max cache size is smaller than the memory needed
> to store all the currently-active tasks and services? We'd end up with
> items disappearing before we have a chance to actually report them, and get
> silent data loss.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/weaveworks/scope/pull/2065#issuecomment-267676268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQOJEGEg12z1oLs4hURfibdxzKB9Qacks5rIuaRgaJpZM4LGXTD>
> .
>

ekimekim commented:
PTAL.
I now use gcache with an expiring LRU, with user-controllable settings (default 1MiB, 1 hour). It doesn't "refresh" the expiry time upon use, so there'll be a spike of requests every hour, bu in practice this shouldn't matter.
I've made tests that exercise the reporter code by mocking the `EcsClient`. The client code has no coverage right now. I think this is acceptable considering the complexity of mocking out the AWS api.

ekimekim commented:
(I've tested manually, and i'm just fixing some linter pedantry right now)

2opremio commented on probe/awsecs/client.go:
```diff
@@ -1,102 +1,212 @@
-		}
+func newECSTask(task *ecs.Task) EcsTask {
+	return EcsTask{
+		TaskARN:           *task.TaskArn,
```
Why the `*`s?

2opremio replied:
Oh, sorry, it's `ecs.Task` not `ECSTask`, my bad.

2opremio commented on probe/awsecs/client.go:
```diff
@@ -1,102 +1,212 @@
+				}(batch)
+				count += len(batch)
+				calls++
+				batch = make([]string, 0, maxServices)
```
I think it's faster and more descriptive to do `batch := batch[:0]`

2opremio replied:
For instance, there will be no allocations

ekimekim replied:
This will cause the next batch to re-use the same in-memory array as the previous batch. Since we haven't made a copy of the underlying array when passing batch to `describeServicesBatch`, only of the slice (which continues pointing to the same underlying array), this causes subsequent appends to batch to modify entries in the previous batch.

Simplified example: https://play.golang.org/p/FfQf7pMqDK

2opremio replied:
Good point, thanks

2opremio commented on probe/awsecs/client.go:
```diff
@@ -1,102 +1,212 @@
+	log.Debugf("Describing ECS services")
 
+	go func() {
+		const maxServices = 10 // How many services we can put in one Describe command
```
Wouldn't all this be equivalent to having a channel with a buffer of >10? Instead of calling `describeServicesBatch()` in the background, you do it in the foreground and let the channel do the buffering for you.

ekimekim replied:
It's not immediately apparent to me how you'd implement this. How do you block until a channel is full, or has a certain number of elements?
Also, it was an intentional choice to do each describe in the background so that many describe operations could be executing concurrently (my assumption is that any local processing is much faster than any network call)

2opremio replied:
Thanks for the clarifications.

>  it was an intentional choice to do each describe in the background so that many describe operations could be executing concurrently (my assumption is that any local processing is much faster than any network call)

Is it worth the compilation? 

Also, is waiting for 10 services really the right thing? 

The channel needs to explicitly closed by the writer in order to "flush" the services so I don't really see the advantage of using a channel for input. You could pass all the services directly (no input channel), chunk them and spawn go routines. Simpler and more explicit.

ekimekim replied:
> The channel needs to explicitly closed by the writer in order to "flush" the services so I don't really see the advantage of using a channel for input.

The intent was that, with a `ListServices` api call returning pages over time, we could start describing the first 10 before the `ListServices` call was finished (the flush behaviour is only for the last, partial batch). But you're right, it's a huge complication for a negligible gain. Will refactor.

I do still think doing the describes concurrently is worthwhile though. If you have, say, 50 services, and a request takes, say, 100ms round trip, then 5 concurrent requests vs 5 sequential ones is the difference between 100ms and 500ms total wall time.

2opremio replied:
I agree. Thanks!

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +216,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+			continue
+		}
+		task := taskRaw.(EcsTask)
+		if !strings.HasPrefix(task.StartedBy, servicePrefix) {
```
Would it make sense to add a field in `EcsTask`describing wether it's a service in `newECSTask`?

ekimekim replied:
My instinct is to try to keep the data normalized, ie. just store raw fields and not derived ones. But I could do something like `ecsTask.isServiceManaged()` so that the check itself is in one obvious place (and it's not exactly an expensive check).

ekimekim replied:
done

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +216,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+	return results, unmatched
+}
+
+func (c ecsClientImpl) ensureTasks(taskARNs []string) {
```
`addTasksToCache`? `ensureTasksAreCached`?

ekimekim replied:
did the latter

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +216,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+	for _, taskARN := range taskARNs {
+		// It's possible that tasks could still be missing from the cache if describe tasks failed.
+		// We'll just pretend they don't exist.
+		if taskRaw, err := c.taskCache.Get(taskARN); err == nil {
```
Worth factoring it out in a function (this casting seems to be used at at least a couple of places)? (same for services)

Maybe add a cache interface?

ekimekim replied:
done

2opremio commented on prog/main.go:
```diff
@@ -287,6 +289,8 @@ func main() {
 	// AWS ECS
 	flag.BoolVar(&flags.probe.ecsEnabled, "probe.ecs", false, "Collect ecs-related attributes for containers on this node")
+	flag.IntVar(&flags.probe.ecsCacheSize, "probe.ecs.cache.size", 1024*1024, "Max size of cached info for each ECS cluster")
+	flag.DurationVar(&flags.probe.ecsCacheExpiry, "probe.ecs.cache.expiry", time.Hour, "How long to keep cached ECS info")
```
👍

2opremio commented:
Thanks for changing the cache and for the tests. How and for how long did you test this? (Not that I would had been able to make it simpler, but the code is quite involved)

2opremio commented:
> Ran for about 15h with no rate limiting.

3 clusters, right? With how many instances each?

ekimekim commented:
PTAL

Yes, 3 clusters. Each with the defaults for the socks shop demo, which is 3 instances, about 10 or so services with one task per service.

ekimekim commented:
Note, the code has changed since then, but not in ways that matter to the rate-limiting. And I've since re-tested that it still works, but haven't repeated the 15h test.

2opremio commented on probe/awsecs/client.go:
```diff
@@ -106,45 +239,166 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+		servicesRefreshed[serviceName] = true
+	}
+	close(toDescribe)
+	<-done
```
Is done really needed? AFAICT it's not read on the other end.

ekimekim replied:
This is the read here, we're blocking on it being written to or closed inside describeServices. However, by simplifying describeServices to take in args instead of returning an input chan, we will do away with the need for it.

ekimekim commented:
PTAL

2opremio commented on probe/awsecs/client.go:
```diff
@@ -129,85 +129,67 @@ func (c ecsClientImpl) getCachedService(serviceName string) (EcsService, bool) {
-			calls++
-		}
+	// split into batches
+	batches := make([][]string, 0, len(services)/maxServices+1)
```
I think it would be event simpler to spawn the goroutines directly (factoring the spawning into a function called inside the loop) instead of accumulating the batches in slices of slices.

ekimekim replied:
I considered it, but found it awkward due to needing to repeat the spawning procedure for the special final case, the partial batch.

2opremio replied:
Yep, that's why I mention factoring the spawning into a function. Your call though.

2opremio commented:
LGTM. That last comment is a recommendation only.
